util/pkg/vfs/vaultfs.go

```diff
@@ -0,0 +1,257 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vfs
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+
+	vault "github.com/hashicorp/vault/api"
+	"k8s.io/klog"
+)
+
+type VaultPath struct {
+	vaultClient *vault.Client
+	mountPoint  string
+	path        string
+}
+
+var _ Path = &VaultPath{}
+
+func newVaultPath(client *vault.Client, path string) (*VaultPath, error) {
+	path = strings.TrimPrefix(path, "/")
+	dirs := strings.SplitN(path, "/", 2)
+	if len(dirs) != 2 {
+		return nil, fmt.Errorf("vault path must have both a mount point and a path. Got: %s", path)
+	}
+
+	if client == nil {
+		return nil, fmt.Errorf("vault paths needs to have a vault client")
+
+	}
+
+	return &VaultPath{
+		vaultClient: client,
+		mountPoint:  dirs[0],
+		path:        dirs[1],
+	}, nil
+}
+
+func (p *VaultPath) WriteFile(data io.ReadSeeker, acl ACL) error {
+	klog.V(4).Infof("Writing file %q", p)
+
+	file, err := encodeData(data)
+	if err != nil {
+		return err
+	}
+
+	payload := map[string]interface{}{
+		"data": map[string]interface{}{
+			"file": file,
+		},
+	}
+
+	_, err = p.vaultClient.Logical().Write(p.dataPath(), payload)
+	return err
+}
+
+func (p *VaultPath) CreateFile(data io.ReadSeeker, acl ACL) error {
+	file, _ := p.ReadFile()
+	if file == nil {
+		return p.WriteFile(data, acl)
+	} else {
+		return fmt.Errorf("file already exists: %v", p.path)
+	}
+}
+
+func (p *VaultPath) ReadFile() ([]byte, error) {
+	secret, err := p.vaultClient.Logical().Read(p.dataPath())
+	if err != nil {
+		return nil, err
+	}
+
+	if secret == nil || secret.Data["data"] == nil {
+		return nil, os.ErrNotExist
+	}
+
+	data := secret.Data["data"].(map[string]interface{})
+
+	encodedString := data["file"].(string)
+	return base64.StdEncoding.DecodeString(encodedString)
+}
+
+func (p *VaultPath) Remove() error {
+	klog.V(8).Infof("removing file %s", p)
+
+	data := map[string][]string{
+		"versions": {"1"},
```

Comment:
This only deletes version 1? What if the data got updated?

Reply:
It doesn't matter. It is more to satisfy the requirement of the destroy method. After the secret metadata is also deleted, the secret is gone.

Reply:
Comment resolved. I don't have detailed knowledge of the Vault API.